### PR TITLE
Emit metric for various mutable state stats

### PR DESCRIPTION
**Pull request by samarabbas**
MutableState support for computing following metric:
1) Total size of mutableState
2) Execution info size within mutable state
3) Activity info size within mutable state
4) Timer info size within mutable state
5) Child info size within mutable state
6) Signal info size within mutable state
7) Buffered events size within mutable state
8) Buffered replication tasks size within mutable state
9) Number of activity infos within mutable state
10) Number of timers within mutable state
11) Number of child executions within mutable state
12) Number of signals within mutable state
13) Number of requestCancels within mutable state
14) Number of buffered events within mutable state
15) Number of buffered replication tasks within mutable state

Session update support for following metric:
1) TotalSize of mutable state
2) Execution info size for session update
3) Activity info size for session update
4) Timer info size for session update
5) Child info size for session update
6) Signal info size for session update
7) Buffered events size for session update
8) Buffered replication tasks size for session update
9) Number of updated activity infos for session update
10) Number of updated timer infos for session update
11) Number of updated child infos for session update
12) Number of updated signal infos for session update
13) Number of updated requestCancel infos for session update
14) Number of deleted activity infos for session update
15) Number of deleted timer infos for session update
16) Number of deleted child infos for session update
17) Number of deleted signal infos for session update
18) Number of deleted requestCancel infos for session update

Support to emit metric for all mutableState and sessionUpdate stats on each update through workflowContext.

Emit metric on each AppendHistory call for size of update.

Support for keeping track of total history size for workflow execution as part of mutable state.

### CODE REVIEW

File: common/metrics/defs.go [hunk 1]
```diff
@@ -448,6 +452,8 @@ const (
 	ReplicatorTaskHistoryScope
 	// ReplicateHistoryEventsScope is the scope used by historyReplicator API for applying events
 	ReplicateHistoryEventsScope
+	// ShardContextScope is the scope used when by ShardContext for emitting various metric
+	ShardContextScope
 	// ShardInfoScope is the scope used when updating shard info
 	ShardInfoScope
```
Comment (wxing1292) [inline common/metrics/defs.go]: can we combine the ShardContextScope and ShardInfoScope

File: common/metrics/defs.go [hunk 2]
```diff
@@ -951,6 +991,27 @@ var MetricDefs = map[ServiceIdx]map[int]metricDefinition{
 		CacheMissCounter:                             {metricName: "cache-miss", metricType: Counter},
 		AcquireLockFailedCounter:                     {metricName: "acquire-lock-failed", metricType: Counter},
 		WorkflowContextCleared:                       {metricName: "workflow-context-cleared", metricType: Counter},
+		HistorySize:                                  {metricName: "history-size", metricType: Timer},
```
Comment (wxing1292) [inline common/metrics/defs.go]: i think the counter metrics type are all named *Counter, so should be the timers?

File: service/history/shardContext.go [hunk 3]
```diff
@@ -562,11 +568,11 @@ func (s *shardContextImpl) AppendHistoryEvents(request *persistence.AppendHistor
 		if _, ok := err0.(*persistence.ConditionFailedError); ok {
 			// Inserting a new event failed, lets try to overwrite the tail
 			request.Overwrite = true
-			return s.historyMgr.AppendHistoryEvents(request)
+			return size, s.historyMgr.AppendHistoryEvents(request)
 		}
 	}
 
-	return err0
+	return 0, err0
```
Comment (vancexu) [inline service/history/shardContext.go]: Why return `0` not `size` when no error?

Comment (samarabbas) [reply service/history/shardContext.go]: Good catch.  Fixed.

**Pull request merged**